File: backend/app/routes/merchant.py
from fastapi import APIRouter, Request
from psycopg import Connection

from ..db import all_rows, audit, one, set_clause
from ..deps import Db, Merchant
from ..errors import bad_request, conflict, forbidden
from ..http import success
from ..schemas import BusinessEventIn, BusinessPatch, CouponIn, CouponRedeemIn, CouponReverseIn
from ..security import hash_token
from .admin_core import created
from .p2_admin import insert_coupon, redeem_issue


router = APIRouter()


def owned_business(connection: Connection, business_id: str, user: dict, *, approved: bool = False) -> dict:
    row = one(connection, """SELECT fb.*,b.name,b.registration_no FROM festival_businesses fb JOIN businesses b ON b.id=fb.business_id
        JOIN festivals f ON f.id=fb.festival_id WHERE fb.id=%s AND fb.owner_membership_id=%s AND f.organization_id=%s""",
        (business_id, user["membership_id"], user["organization_id"]))
    if not row:
        raise forbidden("BUSINESS_SCOPE_DENIED", "본인 업체만 접근할 수 있습니다.")
    if approved and row["participation_status"] != "APPROVED":
        raise bad_request("BUSINESS_NOT_APPROVED", "승인된 참여업체만 사용할 수 있습니다.")
    return row


@router.get("/merchant/businesses")
def businesses(request: Request, user: Merchant, connection: Db):
    rows = all_rows(connection, """SELECT DISTINCT ON (fb.id) fb.*,b.name,b.registration_no,b.address,bo.booth_no,bo.area_id
        FROM festival_businesses fb JOIN businesses b ON b.id=fb.business_id LEFT JOIN booths bo ON bo.festival_business_id=fb.id
        WHERE fb.owner_membership_id=%s ORDER BY fb.id,bo.booth_no""", (user["membership_id"],))
    rows.sort(key=lambda row: row["updated_at"], reverse=True)
    return success(request, rows)


@router.patch("/merchant/businesses/{business_id}")
def update_business(business_id: str, body: BusinessPatch, request: Request, user: Merchant, connection: Db):
    before = owned_business(connection, business_id, user)
    if before["participation_status"] not in {"DRAFT", "SUBMITTED", "REJECTED", "APPROVED"}:
        raise bad_request("INVALID_STATE_TRANSITION", "수정 가능한 업체 상태가 아닙니다.")
    if body.name is not None:
        connection.execute("UPDATE businesses SET name=%s,updated_at=now() WHERE id=%s", (body.name, before["business_id"]))
    # 업체명만 바꾼 요청도 재검수 대상이라 festival_businesses는 항상 갱신한다.
    values = body.model_dump(exclude_none=True, exclude={"name", "version"})
    if not values and body.name is None:
        raise bad_request("VALIDATION_ERROR", "변경할 값이 없습니다.")
    clause, params = set_clause(values) if values else ("", [])
    row = one(connection, f"""UPDATE festival_businesses SET {clause + ',' if clause else ''}participation_status='SUBMITTED',
        review_comment=NULL,version=version+1,updated_at=now() WHERE id=%s AND version=%s RETURNING *""", [*params, business_id, body.version])
    if not row:
        raise conflict("RESOURCE_VERSION_CONFLICT", "업체 정보가 이미 변경되었습니다.")
    audit(connection, festival_id=str(row["festival_id"]), actor_id=str(user["id"]), action="SUBMIT", resource_type="FESTIVAL_BUSINESS",
          resource_id=business_id, before_data=before, after_data=row, request_id=request.state.request_id)
    return success(request, row)


@router.post("/merchant/businesses/{business_id}/coupons", status_code=201)
def create_coupon(business_id: str, body: CouponIn, request: Request, user: Merchant, connection: Db):
    business = owned_business(connection, business_id, user, approved=True)
    row = insert_coupon(connection, business_id, body, user["id"])
    created(connection, request, user, str(business["festival_id"]), "COUPON", row)
    return success(request, row)


@router.post("/merchant/coupon-issues/{issue_id}/redeem")
def redeem_coupon(issue_id: str, body: CouponRedeemIn, request: Request, user: Merchant, connection: Db):
    issue = one(connection, """SELECT ci.*,(ci.expires_at<=now()) AS expired,c.festival_business_id,c.name,fb.festival_id
        FROM coupon_issues ci JOIN coupons c ON c.id=ci.coupon_id JOIN festival_businesses fb ON fb.id=c.festival_business_id
        WHERE ci.id=%s AND fb.owner_membership_id=%s FOR UPDATE OF ci""", (issue_id, user["membership_id"]))
    if not issue:
        raise forbidden("BUSINESS_SCOPE_DENIED", "본인 업체의 쿠폰만 처리할 수 있습니다.")
    if issue["issue_token_hash"] != hash_token(body.issue_token):
        raise bad_request("INVALID_COUPON_TOKEN", "쿠폰 토큰이 일치하지 않습니다.")
    return success(request, redeem_issue(connection, request, issue, user["id"], str(issue["festival_id"])))


@router.post("/merchant/coupon-redemptions/{redemption_id}/reverse")
def reverse_coupon(redemption_id: str, body: CouponReverseIn, request: Request, user: Merchant, connection: Db):
    redemption = one(connection, """SELECT cr.*,fb.festival_id FROM coupon_redemptions cr JOIN festival_businesses fb ON fb.id=cr.festival_business_id
        WHERE cr.id=%s AND fb.owner_membership_id=%s FOR UPDATE OF cr""", (redemption_id, user["membership_id"]))
    if not redemption:
        raise forbidden("BUSINESS_SCOPE_DENIED", "본인 업체의 쿠폰 사용만 취소할 수 있습니다.")
    if redemption["status"] != "REDEEMED":
        raise conflict("INVALID_COUPON_STATUS", "이미 취소된 사용 내역입니다.")
    row = one(connection, "UPDATE coupon_redemptions SET status='REVERSED',reversed_at=now(),reversal_reason=%s WHERE id=%s RETURNING *", (body.reason, redemption_id))
    connection.execute("UPDATE coupon_issues SET status=CASE WHEN expires_at>now() THEN 'ISSUED' ELSE 'EXPIRED' END WHERE id=%s", (redemption["coupon_issue_id"],))
    audit(connection, festival_id=str(redemption["festival_id"]), actor_id=str(user["id"]), action="REVERSE",
          resource_type="COUPON_REDEMPTION", resource_id=redemption_id, before_data=redemption, after_data=row,
          request_id=request.state.request_id)
    return success(request, row)


@router.post("/merchant/businesses/{business_id}/events", status_code=201)
def record_business_event(business_id: str, body: BusinessEventIn, request: Request, user: Merchant, connection: Db):
    business = owned_business(connection, business_id, user, approved=True)
    if body.event_type == "SALE" and body.sales_amount is None:
        raise bad_request("VALIDATION_ERROR", "매출 이벤트에는 금액이 필요합니다.")
    row = one(connection, "INSERT INTO business_events(festival_business_id,event_type,sales_amount,source) VALUES(%s,%s,%s,%s) RETURNING *",
        (business_id, body.event_type, body.sales_amount, body.source))
    created(connection, request, user, str(business["festival_id"]), "BUSINESS_EVENT", row)
    return success(request, row)


@router.get("/merchant/businesses/{business_id}/performance")
def performance(business_id: str, request: Request, user: Merchant, connection: Db):
    owned_business(connection, business_id, user)
    metrics = all_rows(connection, """SELECT event_type,count(*)::int AS count,coalesce(sum(sales_amount),0) AS sales_amount
        FROM business_events WHERE festival_business_id=%s GROUP BY event_type ORDER BY event_type""", (business_id,))
    coupons = one(connection, """SELECT count(ci.id)::int AS issued,count(cr.id) FILTER(WHERE cr.status='REDEEMED')::int AS redeemed
        FROM coupons c LEFT JOIN coupon_issues ci ON ci.coupon_id=c.id LEFT JOIN coupon_redemptions cr ON cr.coupon_issue_id=ci.id
        WHERE c.festival_business_id=%s""", (business_id,))
    return success(request, {"events": metrics, "coupons": coupons})

File: src/features/merchant.ts
import { adminApi, json } from "@/shared/lib/api";
import type { NewCoupon } from "@/features/business-admin";
import { uniqueById } from "@/shared/lib/utils";

export interface MenuItem {
  name: string;
  price: number;
}

export interface MerchantBusiness {
  id: string;
  festivalId: string;
  name: string;
  registrationNo: string;
  category: string;
  description: string | null;
  /** 방문객 업체 목록·상세에 그대로 나가는 값들. 화면에 편집란이 없어 등록 후 고칠 수 없었다. */
  menu: MenuItem[];
  operatingHours: Record<string, string>;
  accessibility: Record<string, boolean>;
  participationStatus: "DRAFT" | "SUBMITTED" | "APPROVED" | "REJECTED";
  reviewComment: string | null;
  boothNo: string | null;
  version: number;
}

export async function fetchMyBusinesses() {
  const rows = await adminApi<MerchantBusiness[]>("/merchant/businesses");
  // 부스가 여러 개면 업체가 부스 수만큼 중복돼 온다. 업체 단위 화면이라 대표 부스만 남긴다.
  return uniqueById(rows);
}

export interface BusinessUpdate {
  businessId: string;
  version: number;
  name?: string;
  category?: string;
  description?: string;
  menu?: MenuItem[];
  operatingHours?: Record<string, string>;
  accessibility?: Record<string, boolean>;
}

/** 수정하면 참여 상태가 SUBMITTED로 돌아가 재검수를 받는다. */
export function updateMyBusiness({ businessId, ...body }: BusinessUpdate) {
  return adminApi(`/merchant/businesses/${businessId}`, json("PATCH", body));
}

/** 상인 콘솔의 쿠폰 발행 입력. 운영자 콘솔과 본문이 같아 한 곳(business-admin)만 정의한다. */
export type MerchantCoupon = NewCoupon;

export function createMerchantCoupon({ businessId, ...body }: MerchantCoupon & { businessId: string }) {
  return adminApi(`/merchant/businesses/${businessId}/coupons`, json("POST", body));
}

/** 방문객이 보여준 쿠폰 발급 ID와 토큰을 대조해 사용 처리한다. */
export function redeemCoupon({ issueId, issueToken }: { issueId: string; issueToken: string }) {
  return adminApi<{ id: string; status: string }>(`/merchant/coupon-issues/${issueId}/redeem`, json("POST", { issueToken }));
}

export function reverseRedemption({ redemptionId, reason }: { redemptionId: string; reason: string }) {
  return adminApi(`/merchant/coupon-redemptions/${redemptionId}/reverse`, json("POST", { reason }));
}

export function recordBusinessEvent({ businessId, ...body }: { businessId: string; eventType: "VISIT" | "SALE"; salesAmount?: number }) {
  return adminApi(`/merchant/businesses/${businessId}/events`, json("POST", body));
}

export interface MerchantPerformance {
  events: Array<{ eventType: string; count: number; salesAmount: number }>;
  coupons: { issued: number; redeemed: number };
}

export function fetchPerformance(businessId: string) {
  return adminApi<MerchantPerformance>(`/merchant/businesses/${businessId}/performance`);
}
